Extension/src/LanguageServer/clientCollection.ts

```diff
@@ -215,3 +220,69 @@ export class ClientCollection {
         return cpptools.DefaultClient.stopLanguageClient();
     }
 }
+
+interface TimeStampSequence {
+    activationTime: number; // when the file appears in the editor. Defined for both "cold/warm" start cases.
+    setupTime: number; // when the Intellisense_client constructor is completed
+    updateRangeTime: number; // when publishDiagnostics & provideSemanticTokens is completed
+    totalTime: number;
+}
+
+export class TimeTelemetryCollector {
+
+    private cachedTimeStamps: Map<string, any> = new Map<string, any>(); // a map of uri's string to TimeStampSequence
+    private extensionStartTime: number; // when the extension starts to activate.
+    private firstFile: number; // when the extension is activated. Defined only for "cold" start cases.
+
+    private getTimeStamp(uri: string) {
+        return this.cachedTimeStamps.get(uri) ? this.cachedTimeStamps.get(uri) :
+            { activationTime: 0, setupTime: 0, updateRangeTime: 0, totalTime: 0 };
+    }
+
+    public clear() {
+        console.log("clearing timestamp log");
+        this.cachedTimeStamps.clear();
+    }
+
+    constructor() {
+        this.extensionStartTime = new Date().getTime();
+        this.firstFile = 0;
+    }
+
+    public setFirstFile() {
+        if (!this.firstFile){
+            this.firstFile = new Date().getTime();
+        }
+        Telemetry.logLanguageServerEvent("firstFile", { "firstFile": (this.firstFile - this.extensionStartTime).toString() });
```

Comment:
I don't think there should be a separate event for "firstFile".  Instead, this should be a bool on the openFile event.